### PR TITLE
ENG-7273  - ADV Retrofit Refactor

### PR DESCRIPTION
**Pull request by kevinsites**

### CODE REVIEW

File: NeuroID/src/advancedDeviceLib/java/com/neuroid/tracker/service/NIDAdvancedDeviceApiService.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.neuroid.tracker.service
+
+import okhttp3.ResponseBody
+import retrofit2.Call
+import retrofit2.http.GET
+import retrofit2.http.Path
+
+interface NIDAdvancedDeviceApiService {
+    @GET("/a/{key}")
+    fun getADVKey(@Path("key") key: String): Call<ResponseBody>
```
Comment (kishisaka-nid) [inline NeuroID/src/advancedDeviceLib/java/com/neuroid/tracker/service/NIDAdvancedDeviceApiService.kt:10]: you can replace <ResponserBody> with a data model <AVDKeyData>. The ADVKeyData can then be a data class with annotation indicating type and name of the keys in the model from the returning JSON. 

`data class ADVKeyNetworkResponse(@field: Json(name="status")var status: String, @field: Json(name="key")var key: String)`

This the execute will deserialize the response to a ADVKeyNetworkResponse and you can dump the deserializing code in lines 64-79 in the NIDAdvancedDeviceNetworkService.retryRequests().

Comment (kevinsites) [reply NeuroID/src/advancedDeviceLib/java/com/neuroid/tracker/service/NIDAdvancedDeviceApiService.kt]: Updated!

File: NeuroID/src/advancedDeviceLib/java/com/neuroid/tracker/service/AdvancedDeviceIDManagerService.kt [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    private val context:Context,
+    private val logger:NIDLogWrapper,
+    private val sharedPrefs: NIDSharedPrefsDefaults,
+    private val dataStoreManager: NIDDataStoreManager,
+    private val advNetworkService: ADVNetworkService,
+    private val fpjsClient: FingerprintJS? // only for testing purposes, need to create in real time to pass NID Key
+): AdvancedDeviceIDManagerService {
+    companion object {
+        internal val NID_RID = "NID_RID_KEY"
+        internal val defaultCacheValue = "{\"key\":\"NO_KEY\", \"exp\":0}"
```
Comment (quacke-nid) [inline NeuroID/src/advancedDeviceLib/java/com/neuroid/tracker/service/AdvancedDeviceIDManagerService.kt:37]: Can this JSON be generated?

File: NeuroID/src/advancedDeviceLib/java/com/neuroid/tracker/service/AdvancedDeviceIDManagerService.kt [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                        endpointUrl = remoteUrl
+                    )
+                )
+        }
+
+        val remoteIDJob = CoroutineScope(Dispatchers.IO).launch {
+            var maxRetryCount = NIDAdvancedDeviceNetworkService.RETRY_COUNT
+            var retryCount = 0
+
+            while (retryCount < maxRetryCount) {
```
Comment (quacke-nid) [inline NeuroID/src/advancedDeviceLib/java/com/neuroid/tracker/service/AdvancedDeviceIDManagerService.kt:115]: Restructure loop to use intrinsic index:

```suggestion
for (retryCount in 1..maxRetryCount) {
```

**Review by quacke-nid — APPROVED** [review]

**Review by cometman — APPROVED** [review]

**Review by kishisaka-nid — APPROVED** [review]
pulled, tested, looks good. 🚢

**Pull request merged**